### ISSUE TITLE
Adapt server certificate verification interface so it can be used by platform verifiers

### ISSUE THREAD

**Issue by jsha** [rustls/rustls#850]
In https://github.com/rustls/rustls-native-certs/issues/25 we discussed the desirability of using the platform verifier on macOS and Windows instead of trying to extract a set of roots, because distrust of certain roots is implemented in the platform verifier.

[The platform verifiers perform blocking I/O]( https://github.com/rustls/rustls-native-certs/issues/25#issuecomment-943358491), so the current ServerCertVerifier interface is not appropriate, since it is likely to be called on a thread doing async I/O, where we don't want to block the whole thread.

#787 is an example of solving a similar problem for the server side, but that has the advantage that accepting a ClientHello can happen before any TLS work is done. Server certificate verifier happens in the middle of the handshake.

**Comment by Ralith** [comment]
> Server certificate verifier happens in the middle of the handshake.

I think the same principle stands: create an object for an outgoing connection, drive I/O on it until it reports certificate availability, then perform certificate verification and pass the result into a constructor for the next state.

**Comment by ctz** [comment]
I think maybe we'll want to solve a few issues with the same/similar mechanism:

- asynchronous server certificate validation by client (the primary problem)
- asynchronous client certificate validation by server (#468)
- asynchronous handshake authentication signature by server (mention in passing in #390)
- asynchronous handshake authentication signature by client (ditto)
- asynchronous session storage (#264 -- though i'm not totally convinced, i would expect any scalable distributed session storage implementation would serve from a complete store in memory, with a weakly consistent cache coherence protocol underneath?)

**Comment by jsha** [comment]
What about a callback-based approach?

```rust
pub trait ServerCertVerifier  {
    fn verify_server_cert(
        &self,
        end_entity: &Certificate,
        intermediates: &[Certificate],
        server_name: &rustls::ServerName,
        _scts: &mut dyn Iterator<Item = &[u8]>,
        ocsp_response: &[u8],
        _now: SystemTime,
        callback: impl FnOnce(Result<ServerCertVerified, Error>),
    )
```

When this is called, the user code could choose to do one of:

 - do the validation synchronously, call the callback immediately and return, or
 - do the validation on a threadpool and arrange to call the callback when done.

The ClientConnection would keep an internal bool, `awaiting_verification`. It would be set when the verification was requested and cleared when the callback is called. While the bit was set:

 - `process_new_packets` would continue succeeding but would not actually process additional packets
 - `Reader::read` would return WouldBlock.
 - wants_read would return false
 
There are some tricky lifetime issues. The callback would need to own something it could use to affect the internal state of the connection, but it can't hold onto a `&mut ClientConnection` longer than the duration of the `process_new_packets` call. Perhaps the internal bool should actually be an `Arc<Mutex<bool>>`, shared with the callback.

**Comment by djc** [comment]
@jsha why would that be better than externalizing the verifier similar to how the `Acceptor` does? Intuitively, the `Acceptor`-based approach seems cleaner than a complex callback setup.

**Comment by jsha** [comment]
> why would that be better than externalizing the verifier similar to how the Acceptor does? Intuitively, the Acceptor-based approach seems cleaner than a complex callback setup.

The Acceptor approach is great, but I don't see how to apply it to server certificate verification (and the other cases @ctz mentioned above). Acceptor is pretty straightforward because it can be created without config, and turned into a Connection.

Applying the same concept to ClientConnection seems like it would split it into three states: ClientConnectionBeforeCertReceived, ClientConnectionAwaitingCertVerification, ClientConnection. And each place where we needed to go asynchronous would result in a similar split into multiple types. Am I understanding the `Acceptor`-based approach correctly?

**Comment by djc** [comment]
Yes, I think something like that would be the proposal.

**Comment by Ralith** [comment]
It's worth noting that separate types aren't *necessary*, they're just a lot less error prone due to correctness by construction.

**Comment by jsha** [comment]
Is the idea that all client connections would go through this three-phase process, or only client connections that have a custom verifier, or only clients that have a custom verifier that wants to do blocking I/O?

If we make all clients deal with the three phases, that seems like imposing complexity on the common case for the sake of a less common case. If we only do the three-phase process for connections with a custom verifier, that means there are additional, quite different, paths through the rustls code - which makes testing more complicated.

Sketching this out to clarify in my mind:

ClientConnectionBeforeCertReceived: wants_read(), wants_write(), read_tls(), write_tls() same as today. No read() or write(). No process_new_packets(). New method server_cert(), which returns Option<ClientConnectionAwaitingCertVerification>.
ClientConnectionAwaitingCertVerification: only two methods: info() (returns end_entity, intermediates, server_name, SCTs, etc) and verified() (returns ClientConnection)

How does the three-phase process work with Stream and StreamOwned?

An advantage of the callback approach is that it fits right into the current API design, and doesn't even have to be a breaking change. We could keep the current verify_server_cert method on the trait, and add a verify_server_cert_with_callback with a default implementation that calls verify_server_cert, then immediately calls the callback. That would work the same for all current implementations, and allow new implementations that need to do blocking work.

**Comment by djc** [comment]
While callbacks on threadpools are the way asynchronicity works in C, I think for Rust users we'd want to be able to use async/await (because it's generally much easier to understand than callback code), which doesn't work with the callback syntax. This is exactly why the `Acceptor` API is designed the way it is now.

I guess we could implement a `Future`-based wrapper API that doesn't need to know about the three phases (or we leave that tokio-rustls et al).

For `Stream` and `StreamOwned` I suppose we could signal blocking I/O by returning `Err(io::Error::new(io::ErrorKind::WouldBlock, ()))`.

**Comment by briansmith** [comment]
FWIW, I have helped implement a native platform certificate verifier for Rustls and I have also implemented asynchronous certificate verification in Firefox. I agree it is important that the verifier interface be made asynchronous for the reasons given.

Rustls currently parses the raw certificate data, stapled OCSP responses, and SCTs out of the handshake messages and stores them in `Vec<u8>`s until the handshake is over. Then it tries to verify the certificate chain. This is inefficient in terms of memory and in terms of time. 

Instead, it would be better, whenever Rustls parses one of these objects, for Rustls to avoid allocating a `Vec<u8>` and instead ask the verifier to save the object itself. That way, the verifier can store the object in a platform-specific way (e.g. a Windows `HANDLE`).

Also, instead of waiting until the handshake is finished, Rustls needs to give the verifier the opportunity to start the verification sooner, so it can run in parallel with the processing of the rest of the handshake messages.

When verification is finished, the verifier should be notified that it can throw away whatever state is no longer needed.

Instead of providing the calling application with the raw sequence of encoded certificates as they existed in the handshake, Rustls should provide the calling application with the certificate chain that was built and validated by the validator, which could be much different (e.g. if AIA fetching were used to fetch additional certificates, and/or if the TLS handshake messages include extra certificates that aren't needed to build a path to the trust anchor). Before validation is triggered, the certificate verifier needs to know whether it should try to save this data (the application wants to reserve the right to retrieve it) or if the application would rather save RAM by not saving it.

When storing a session in the session cache, the verifier needs to be able to optionally tell Rustls what time the certificate verification result expires, so that the session cache entry's expiration time can be capped at this time. (Less urgent.)

In the session cache, we should have the ability to store the built certificate chain from the initial verification so that it can be retrieved for any connection, regardless of whether a session was resumed or a new session was created.

Having two separate interfaces, one for server certificate verification and one for client certificate verification, is undesirable. (The `webpki` crate has a similar problem and I'm about to fix it.) There should be a single interface.

With this in mind, I suggest:
* When Rustls encounters the first certificate-related object (X.509 cert, OCSP response, SCT), it asks the verifier to create a new verification state, of type `Box<dyn CertVerification>`, passing it any relevant options (whether to save certificate chain, whether we're verifying a client or a server cert, etc).
* Every time an object is encountered, that object is added to the `CertVerification`. So you could imagine having `add_certificate_status`, `add_certificate`, `add_sct` in `CertVerification`.
* As soon as Rustls has the information for verifying the handshake signature, it should tell the `CertificateVerification` that information, instead of storing it.
* As soon as Rustls has reached the point where no more such objects are possible to be encountered in the handshake, it should tell the verifier so that it can start verifying, potentially asynchronously.
* When Rustls reaches the point at which it currently does certificate verification (processing `Finished`), it will poll the `CertVerification` for its status, and block the handshake until the status is available.
* When Rustls canceled a handshake due to error or whatever, it needs to cancel whatever verification is in progress.
* When deframing/parsing the `Certificate` message, Rustls shouldn't allocate a giant buffer for storing the entire message. Instead, there should be an optimization in the deframer that allows processing the `Certificate` message in parts. (This is something to be done later, but it should be kept in mind so as to understand all the above considerations.)

Note that this would allow the same parallelization of the TLS handshake and certificate verification that is in Firefox, and in fact would allow even better optimizations than are in any other public implementation I know of.

**Comment by ctz** [comment]
I think as time goes on, my preference to address this is to "bite the bullet" and support async as a first class feature in rustls. That would mean having async versions of the various extension traits listed above, and an async versions of the . Potentially it would also simplify the planned "write through" feature to reduce the buffering rustls's IO-agnostic API currently requires.

There are a bunch of negatives with doing this, though. One is basically the "colouring" problem that most async systems have -- would we want to support both the current simple-synchronous-but-buffering API, plus the async one? I'd really like to understand the cost to callers of removing the synchronous APIs (both traits and top-level API). The other is that async functions are not supported in traits, so we'd have to evaluate a dependency which fixes that (ie, async_trait?).

**Comment by taikulawo** [comment]
> I think as time goes on, my preference to address this is to "bite the bullet" and support async as a first class feature in rustls. That would mean having async versions of the various extension traits listed above, and an async versions of the top-level API. Potentially it would also simplify the planned "write through" feature to reduce the buffering rustls's IO-agnostic API currently requires.
> 
> There are a bunch of negatives with doing this, though. One is basically the "colouring" problem that most async systems have -- would we want to support both the current simple-synchronous-but-buffering API, plus the async one? I'd really like to understand the cost to callers of removing the synchronous APIs (both traits and top-level API). The other is that async functions are not supported in traits, so we'd have to evaluate a dependency which fixes that (ie, async_trait?).

can we just add poll-waker based api? we can use same poll api underlying,and expose sync api(just return poll::ready with data) to fit current public rustls api. if user want to make api async, just create their own waker logic

**Comment by yorickpeterse** [comment]
For https://github.com/inko-lang/inko I'm working on TLS support to resolve https://github.com/inko-lang/inko/issues/329. I'm using rustls-platform-verifier to handle certificate verification. On macOS (and potentially Windows) this _may_ result in certificate validation taking longer than desired ([see here](https://github.com/rustls/rustls-native-certs/issues/25#issuecomment-943358491)). In case of an async runtime, that in turn may result in OS threads being blocked.

To work around this, I'd need to be able to do two things:

1. Install some sort of hook that runs around the verification system call that might be slow
2. Access some state, in my case the currently running task

Using those two I'd then be able to use Inko's runtime support for handling (potentially) slow system calls.

This isn't possible today as the verification traits/implementations don't allow passing in extra data that can change on a per verification basis (e.g. the currently running process). Storing this data in e.g. a `ServerConfig` wouldn't work either, as in my case the data could be moved between tasks.

To work around this, I had to fork/vendor rustls-platform-verifier and patch it, using a thread-local to access the currently running task. The result is [this](https://github.com/inko-lang/inko/blob/ceccf17d691fe2454ada5602ff7480d77323a9c5/rt/src/rustls_platform_verifier/verification/apple.rs#L125-L130).

To make this nicer, I think I'd need:

- The ability to either "wrap" an existing verifier instead of having to re-implement it, so I could use rustls-platform-verifier as-is
- The ability to inject extra state into the verifiers when running the verification, such that it can be different every time

What this looks like API wise in rustls, I'm not sure of :confused:
